File: 1.creational/Factory.java
//  Factory Pattern:
//      It creates pile of objects as a production machine (in other terms)
//      Only you should pass the necessary type of what you want it to create
   
import java.lang.System; 

public class Factory { 

    static SaladMakingFactory saladMaker = new SaladMakingFactory();

    public static void main(String[] args) {
        System.out.println("\n Factory pattern exercise begins!!"); 
        System.out.println("\n ... \n ... \n ...  \n ...  \n ... ");
        saladMaker.mixIt(saladMaker.makeItRightNow(SaladType.VegetableSalad));
    }

    interface Ingredient {
        void smells();
        void tastes();
    }
     
    interface Salad {
        Ingredient[] makeItRightNow(Factory.SaladType type);
        void mixIt(Factory.Ingredient[] choppedInBowl);
    }
 
    enum SaladType {
        FruitSalad,
        VegetableSalad,
        OtherType,
        None
    }
}

class SaladMakingFactory implements Factory.Salad {   
    @Override
    public Factory.Ingredient[] makeItRightNow(Factory.SaladType type) { 
        switch(type) {
            case FruitSalad:  
                return new Factory.Ingredient[] {
                    new Fruit("Banana"),
                    new Fruit("Strawberry"),
                    new Fruit("Kiwi"),
                    new Fruit("Chocolate Sauce"),
                }; 
            case VegetableSalad: 
                return new Factory.Ingredient[] {
                    new Vegetable("Tomato"),
                    new Vegetable("Sweet Pepper"),
                    new AnySubstance("Salt"),
                    new AnySubstance("Dib Roman"),
                    new AnySubstance("Sunflower Seed Oil"),
                };
            case OtherType: 
            case None:  
                return null;
        }
        return null;
    }

    @Override
    public void mixIt(Factory.Ingredient[] choppedInBowl) {
        if (choppedInBowl == null) { 
            System.out.print("\n Sorry! \n Nothing exists in the bowl!");
        } else {
            for (Factory.Ingredient i : choppedInBowl) {
                i.smells();
                i.tastes();
            }
            System.out.println("\n ... \n ... \n ...  \n ...  \n ... ");
            System.out.print("\n Done!");
            System.out.print("\n Ready to enjoy! \n \n");
        }
    }
}

class Vegetable implements Factory.Ingredient {
  
    String name; 
    Vegetable(String name) {
        this.name = name;
    }

    @Override
    public void smells() {
        System.out.printf("\n Any %s has a smell", this.name);
    }

    @Override
    public void tastes() {
        System.out.printf("\n Any %s has a taste", this.name);
    }
}

class Fruit implements Factory.Ingredient {
  
    String name; 
    Fruit(String name) {
        this.name = name;
    }
  
    @Override
    public void smells() {
        System.out.printf("\n I dont take the smell on %s generally", this.name);
    }

    @Override
    public void tastes() {
        System.out.printf("\n Most fruits tastes good, have you tried %s?", this.name);
    }
}

class AnySubstance implements Factory.Ingredient {
  
    String name; 
    AnySubstance(String name) {
        this.name = name;
    }
  
    @Override
    public void smells() {
        System.out.printf("\n %s may or may not smell!", this.name);
    }

    @Override
    public void tastes() {
        System.out.printf("\n %s donest have to have a taste", this.name);
    }
} 

File: 1.creational/Factory.swift
//
//  Factory.swift
//  patterns
//
//  Created by Developer on 29.09.2019.
//  Copyright © 2019 Developer. All rights reserved.
//  -------------------------------------------------------------------------
//  Factory Pattern:
//      It creates pile of objects as a production machine (in other terms)
//      Only you should pass the necessary type of what you want it to create
//  -------------------------------------------------------------------------


import Foundation

protocol Ingredient {
   func smells()
   func tastes()
}
 
protocol Salad {
    func makeItRightNow(type:SaladType) -> [Ingredient]?
    func mixIt(choppedInBowl:[Ingredient]?)
}

enum SaladType {
   case FruitSalad
   case VegetableSalad
   case OtherType
   case None
}

class SaladMakingFactory: Salad {
    
    func makeItRightNow(type: SaladType) -> [Ingredient]? {
        switch type {
            case .FruitSalad:
                return [
                    Fruit(name: "Banana"),
                    Fruit(name: "Strawberry"),
                    Fruit(name: "Kiwi"),
                    AnySubstance(name: "Chocolate Sauce"),
                    AnySubstance(name: "M&Ms")
                ]
            case .VegetableSalad:
                return [
                    Vegetable(name: "Tomato"),
                    Vegetable(name: "Sweet Pepper"),
                    AnySubstance(name: "Salt"),
                    AnySubstance(name: "Dib Roman"),
                    AnySubstance(name: "Sunflower Seed Oil")
                ]
            case .OtherType:
                return nil
            case .None:
                return nil
        }
    }
    
    func mixIt(choppedInBowl: [Ingredient]?) {
        if let choppedInBowl = choppedInBowl {
            for i in choppedInBowl {
                i.smells()
                i.tastes()
            }
            print("\n ... \n ... \n ...  \n ...  \n ... ");
            print("\n Done!");
            print("\n Ready to enjoy! \n \n");
        }
    }
}

class Vegetable: Ingredient {
    
    var name:String
    
    init(name:String) { self.name = name }
    
    func smells() {
        print("\nAny \(self.name) has a smell")
    }
    
    func tastes() {
         print("\nAny \(self.name) has a taste");
    }
}

class Fruit: Ingredient {
    
    var name:String
    
    init(name:String) { self.name = name }
    
    func smells() {
        print("\nI dont take the smell on \(self.name) generally")
    }
    
    func tastes() {
        print("\nMost fruits tastes good, have you tried \(self.name)?");
    }
}

class AnySubstance: Ingredient {
    
    var name:String
    
    init(name:String) { self.name = name }
    
    func smells() {
        print("\n\(self.name) may or may not smell!")
    }
    
    func tastes() {
        print("\n\(self.name) doesn't have to have a taste")
    }
}

 
//  -------------------------------------------------------------------------
//  -------------------------------------------------------------------------
//                                  TEST NOW
//  -------------------------------------------------------------------------
//  -------------------------------------------------------------------------
 

print("\nFactory pattern exercise begins!!")
print("\n ... \n ... \n ...  \n ...  \n ... ")

let saladMaker = SaladMakingFactory();
saladMaker.mixIt(choppedInBowl: saladMaker.makeItRightNow(type: .VegetableSalad))
